File: packages/plugin-demo-ruby-http/backend.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# CodeExpander 插件后端 - http 模式 (Ruby)
#
# 协议：从环境变量 CODEEXPANDER_BACKEND_PORT 获取端口
#       在 127.0.0.1:PORT 启动 HTTP 服务
#       接收 POST / 请求，body 为 { method, payload }
#       返回 JSON 响应
#
# 进程常驻，空闲约 5 分钟后退出

require 'json'
require 'webrick'

def handle_request(method, payload)
  case method
  when 'run', 'transform'
    text = payload['text'] || ''
    action = payload['action'] || 'upper'
    result = case action
             when 'upper' then text.upcase
             when 'lower' then text.downcase
             when 'reverse' then text.reverse
             else text
             end
    { ok: true, result: result, action: action }
  when 'ping'
    { ok: true, pong: true, mode: 'http' }
  else
    { ok: false, error: "Unknown method: #{method}" }
  end
end

port = ENV['CODEEXPANDER_BACKEND_PORT']
if port.nil? || port.empty?
  $stderr.puts 'CODEEXPANDER_BACKEND_PORT not set'
  exit 1
end

server = WEBrick::HTTPServer.new(Port: port.to_i, BindAddress: '127.0.0.1')
server.mount_proc '/' do |req, res|
  if req.request_method != 'POST' || req.path != '/'
    res.status = 404
    next
  end
  begin
    data = JSON.parse(req.body)
    method = data['method'] || 'run'
    payload = data['payload'] || {}
    response = handle_request(method, payload)
    res['Content-Type'] = 'application/json'
    res.body = JSON.generate(response)
  rescue => e
    res.status = 500
    res['Content-Type'] = 'application/json'
    res.body = JSON.generate(ok: false, error: e.message)
  end
end

trap('INT') { server.shutdown }
trap('TERM') { server.shutdown }
server.start

File: examples/plugin-demo-nodejs-http/backend.js
#!/usr/bin/env node
/**
 * CodeExpander 插件后端 - http 模式
 *
 * 协议：从环境变量 CODEEXPANDER_BACKEND_PORT 获取端口
 *       在 127.0.0.1:PORT 启动 HTTP 服务
 *       接收 POST / 请求，body 为 { method, payload }
 *       返回 JSON 响应
 *
 * 进程常驻，多次调用复用同一进程，空闲约 5 分钟后退出
 */

const http = require("http");

const PORT = process.env.CODEEXPANDER_BACKEND_PORT;
if (!PORT) {
  console.error("CODEEXPANDER_BACKEND_PORT not set");
  process.exit(1);
}

function handleRequest(method, payload) {
  switch (method) {
    case "run":
    case "transform": {
      const { text = "", action = "upper" } = payload;
      let result = text;
      if (action === "upper") result = text.toUpperCase();
      else if (action === "lower") result = text.toLowerCase();
      else if (action === "reverse") result = text.split("").reverse().join("");
      return { ok: true, result, action };
    }
    case "ping":
      return { ok: true, pong: true, mode: "http" };
    default:
      return { ok: false, error: `Unknown method: ${method}` };
  }
}

const server = http.createServer((req, res) => {
  if (req.method !== "POST" || req.url !== "/") {
    res.writeHead(404);
    res.end();
    return;
  }
  let body = "";
  req.on("data", (chunk) => { body += chunk; });
  req.on("end", () => {
    try {
      const { method = "run", payload = {} } = JSON.parse(body);
      const response = handleRequest(method, payload);
      res.writeHead(200, { "Content-Type": "application/json" });
      res.end(JSON.stringify(response));
    } catch (err) {
      res.writeHead(500, { "Content-Type": "application/json" });
      res.end(JSON.stringify({ ok: false, error: String(err.message) }));
    }
  });
});

server.listen(Number(PORT), "127.0.0.1", () => {
  // 保持进程存活，CodeExpander 会复用此进程
});
